### PR TITLE
Class "llvm:StringRef" has no member 'Startswith'

### PR DESCRIPTION
**Pull request by Darkhost1999**
Following this example:
https://github.com/RPCS3/rpcs3/blob/master/Utilities%2FJITLLVM.cpp#L73

I resolved:
```
Class "llvm:StringRef" has no member 'Startswith'
if (brand.startswith("AMD"))
```
and
```
Class "llvm:StringRef" has no member 'Startswith'
else if (brand.startswith("VirtualApple"))
```
observed when checking out
LLVM 19.1.0-rc2

This resolves 1 problem with compilation on later versions of LLVM

### CODE REVIEW

**Comment by Nekotekina** [comment]
I think you want to call starts_with for std::string, not StringRef

**Comment by Nekotekina** [comment]
I mean, your PR can probably cause other compatibility issues (with older LLVM versions?)

**Comment by elad335** [comment]
maybe use some ifdefs?

**Comment by Megamouse** [comment]
It seems to build with our llvm version. I don't see a reason why it would need to work with even older versions

**Comment by kd-11** [comment]
> It seems to build with our llvm version. I don't see a reason why it would need to work with even older versions

I think we support both 16 and 17 at the moment outside official CI. This is useful for LTS systems using system llvm versions which can be a bit dated.
However, it may be time to bump llvm up again. Ubuntu 24.04 LTS already has been around for a while, enough for downstream distros to update. I can look into that once I'm done with arm64 stuff.

**Comment by Megamouse** [comment]
I don't think Ubuntu 24.04 is officially rolled out yet

**Comment by Nekotekina** [comment]
`std::string_view(string).starts_with` might work as StringRef has a conversion operator

**Comment by Megamouse** [comment]
Btw, you shouldn't really try to fix any issues with llvm 19 if there isn't even an official release.
For all we know this might change still.

**Comment by Darkhost1999** [comment]
I see Ubuntu Noble (24.04) enabled in LLVM news on March 22nd, 2024.
LLVM 18.1.8 tagged June 18th, 2024.
And RPCS3 experiences the same JITLLVM compilation issue on release tag 18.1.8
It isn't the only issue with compilation on any of the future tags. Just 1 observed.

**Comment by oltolm** [comment]
```
[279/868] Building CXX object rpcs3/Emu/CMakeFiles/rpcs3_emu.dir/__/__/Utilities/JITLLVM.cpp.obj
C:/src/rpcs3/Utilities/JITLLVM.cpp:730:26: warning: object backing the pointer will be destroyed at the end of the full-expression [-Wdangling-gsl]
  730 |         llvm::StringRef brand = utils::get_cpu_brand();
      |                                 ^~~~~~~~~~~~~~~~~~~~~~
C:/src/rpcs3/Utilities/JITLLVM.cpp:734:12: warning: 'startswith' is deprecated: Use starts_with instead [-Wdeprecated-declarations]
  734 |         if (brand.startswith("AMD"))
      |                   ^~~~~~~~~~
      |                   starts_with
C:/msys64/clang64/include/llvm/ADT/StringRef.h:261:19: note: 'startswith' has been explicitly marked deprecated here
  261 |     [[nodiscard]] LLVM_DEPRECATED(
      |                   ^
C:/msys64/clang64/include/llvm/Support/Compiler.h:155:50: note: expanded from macro 'LLVM_DEPRECATED'
  155 | #define LLVM_DEPRECATED(MSG, FIX) __attribute__((deprecated(MSG, FIX)))
      |                                                  ^
C:/src/rpcs3/Utilities/JITLLVM.cpp:782:17: warning: 'startswith' is deprecated: Use starts_with instead [-Wdeprecated-declarations]
  782 |         else if (brand.startswith("VirtualApple"))
      |                        ^~~~~~~~~~
      |                        starts_with
C:/msys64/clang64/include/llvm/ADT/StringRef.h:261:19: note: 'startswith' has been explicitly marked deprecated here
  261 |     [[nodiscard]] LLVM_DEPRECATED(
      |                   ^
C:/msys64/clang64/include/llvm/Support/Compiler.h:155:50: note: expanded from macro 'LLVM_DEPRECATED'
  155 | #define LLVM_DEPRECATED(MSG, FIX) __attribute__((deprecated(MSG, FIX)))
      |                                                  ^
3 warnings generated.
```

I get these warnings when compiling with clang 18.1.8 against LLVM 18.1.8. `brand` should be a `std::string` and other places changed accordingly because `brand` is a `StringRef` that outlives the string returned by `utils::get_cpu_brand()`.

**Comment by Darkhost1999** [comment]
> C:/src/rpcs3/Utilities/JITLLVM.cpp:730:26: warning: object backing the pointer will be destroyed at the end of the full-expression [-Wdangling-gsl]

This should be resolved and line 762 saying that contains isn't a member of std:string was changed to find to be a member of std:string.

File: Utilities/JITLLVM.cpp [hunk 1]
```diff
@@ -779,7 +779,7 @@ llvm::StringRef fallback_cpu_detection()
 		}
 		return "icelake-client";
 	}
-	else if (brand.startswith("VirtualApple"))
+	else if (std::string_view(brand).starts_with("VirtualApple"))
```
Comment (Megamouse) [inline Utilities/JITLLVM.cpp:782]:
```suggestion
	else if (brand.starts_with("VirtualApple"))
```

File: Utilities/JITLLVM.cpp [hunk 2]
```diff
@@ -759,7 +759,7 @@ llvm::StringRef fallback_cpu_detection()
 				: "znver3";
 		}
 	}
-	else if (brand.contains("Intel"))
+	else if (brand.find("Intel"))
```
Comment (Megamouse) [inline Utilities/JITLLVM.cpp:762]:
```suggestion
	else if (brand.find("Intel") != usz)
```

File: Utilities/JITLLVM.cpp [hunk 3]
```diff
@@ -727,11 +727,11 @@ llvm::StringRef fallback_cpu_detection()
 {
 #if defined (ARCH_X64)
 	// If we got here we either have a very old and outdated CPU or a new CPU that has not been seen by LLVM yet.
-	llvm::StringRef brand = utils::get_cpu_brand();
+	std::string brand = utils::get_cpu_brand();
 	const auto family = utils::get_cpu_family();
 	const auto model = utils::get_cpu_model();
 
-	if (brand.startswith("AMD"))
+	if (std::string_view(brand).starts_with("AMD"))
```
Comment (Megamouse) [inline Utilities/JITLLVM.cpp:734]:
```suggestion
	if (brand.starts_with("AMD"))
```

File: Utilities/JITLLVM.cpp [hunk 4]
```diff
@@ -727,11 +727,11 @@ llvm::StringRef fallback_cpu_detection()
 {
 #if defined (ARCH_X64)
 	// If we got here we either have a very old and outdated CPU or a new CPU that has not been seen by LLVM yet.
-	llvm::StringRef brand = utils::get_cpu_brand();
+	std::string brand = utils::get_cpu_brand();
```
Comment (Megamouse) [inline Utilities/JITLLVM.cpp:730]:
```suggestion
	const std::string brand = utils::get_cpu_brand();
```

**Pull request merged**